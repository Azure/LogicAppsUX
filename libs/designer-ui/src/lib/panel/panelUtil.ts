/* eslint-disable no-param-reassign */
import { getPropertyValue } from '@microsoft-logic-apps/utils';

export enum PanelLocation {
  Left = 'LEFT',
  Right = 'RIGHT',
}

export enum PanelScope {
  AppLevel = 'APP_LEVEL',
  CardLevel = 'CARD_LEVEL',
}

export enum PanelSize {
  Auto = 'auto',
  Small = '300px',
  Medium = '630px',
}

export interface PanelTab {
  name: string;
  title: string;
  description?: string;
  icon?: string;
  enabled?: boolean;
  order: number;
  content: React.ReactElement;
  visibilityPredicate?(): boolean;
}

export interface CommonPanelProps {
  isCollapsed: boolean;
  toggleCollapse: () => void;
  width: string;
}

export function registerTabs(tabsInfo: PanelTab[], registeredTabs: Record<string, PanelTab>): Record<string, PanelTab> {
  tabsInfo.forEach((tabInfo) => {
    registerTab(tabInfo, registeredTabs);
  });

  return { ...registeredTabs };
}

export function registerTab(tabInfo: PanelTab, registeredTabs: Record<string, PanelTab>): Record<string, PanelTab> {
  // Commenting this out, if the tab is already registered, we might want to update it instead of error out
  // We could possibly have a separate update method, but I think this is the simplest way to do it - Riley
  /*
    const intl = getIntl();
    const tabAlreadyRegistered = intl.formatMessage(
      {
        defaultMessage: 'Tab with : {tabname} name is already registered',
        description: 'This is the  message shown in case of an error of a tab already existing with the name.',
      },
      { tabname: tabInfo.name }
    );
    if (getTab(tabInfo.name, registeredTabs)) {
      throw new ValidationException(ValidationErrorCode.UNSPECIFIED, format(tabAlreadyRegistered, tabInfo.name));
    }
  */
  registeredTabs[tabInfo.name.toLowerCase()] = tabInfo;
  return registeredTabs;
}

export function unregisterTab(name: string, registeredTabs: Record<string, PanelTab>): void {
  if (Object.keys(registeredTabs).some((registeredTabName) => registeredTabName === name)) {
    delete registeredTabs[name];
  }
}

export function getTabs(sort: boolean, registeredTabs: Record<string, PanelTab>): PanelTab[] {
  // Get all tabs not specifically defined as not enabled
  const enabledTabs = Object.values(registeredTabs).filter((tab) => tab.enabled !== false);
  return sort ? enabledTabs.sort((a, b) => a.order - b.order) : enabledTabs;
}

export function getTab(name: string, registeredTabs: Record<string, PanelTab>): PanelTab {
  return getPropertyValue(registeredTabs, name);
}
