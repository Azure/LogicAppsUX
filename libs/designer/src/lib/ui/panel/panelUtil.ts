import { getIntl } from '@microsoft-logic-apps/intl';
import { format, getPropertyValue, ValidationErrorCode, ValidationException } from '@microsoft-logic-apps/utils';

export enum PanelLocation {
  Right = 'RIGHT',
}

export enum PanelTabScope {
  AppLevel = 'APP_LEVEL',
  CardLevel = 'CARD_LEVEL',
}

export interface PanelTab {
  name: string;
  title: string;
  description?: string;
  icon?: string;
  enabled?: boolean;
  order: number;
  content: React.ReactElement;
  visibilityPredicate?(): boolean;
}

export function registerTab(tabInfo: PanelTab, registeredTabs: Record<string, PanelTab>): Record<string, PanelTab> {
  const intl = getIntl();
  const tabAlreadyRegistered = intl.formatMessage(
    {
      defaultMessage: 'Tab with : {tabname} name is already registered',
      description: 'This is the  message shown in case of an error of a tab already existing with the name.',
    },
    { tabname: tabInfo.name }
  );
  if (getTab(tabInfo.name, registeredTabs)) {
    throw new ValidationException(ValidationErrorCode.UNSPECIFIED, format(tabAlreadyRegistered, tabInfo.name));
  }
  registeredTabs[tabInfo.name.toLowerCase()] = tabInfo;
  return registeredTabs;
}

export function unregisterTab(name: string, registeredTabs: Record<string, PanelTab>): void {
  if (Object.keys(registeredTabs).some((registeredTabName) => registeredTabName === name)) {
    delete registeredTabs[name];
  }
}

export function deleteAllTabs(registeredTabs: Record<string, PanelTab>): void {
  registeredTabs = {};
}

export function getTabs(sort: boolean, registeredTabs: Record<string, PanelTab>): PanelTab[] {
  if (sort) {
    const tabsSorted = Object.keys(registeredTabs).map((name) => registeredTabs[name]);
    tabsSorted.sort((a, b) => a.order - b.order);
    return tabsSorted;
  } else {
    return Object.keys(registeredTabs).map((name) => registeredTabs[name]);
  }
}

export function getTab(name: string, registeredTabs: Record<string, PanelTab>): PanelTab {
  return getPropertyValue(registeredTabs, name);
}
