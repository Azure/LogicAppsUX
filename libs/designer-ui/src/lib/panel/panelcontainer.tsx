import type { ILayerProps } from '@fluentui/react';
import { MessageBar, MessageBarType } from '@fluentui/react';
import { Button, Divider, mergeClasses, OverlayDrawer, Spinner } from '@fluentui/react-components';
import { ChevronDoubleRightFilled } from '@fluentui/react-icons';
import { useCallback, useEffect, useRef } from 'react';
import { useIntl } from 'react-intl';
import { EmptyContent } from '../card/emptycontent';
import type { PageActionTelemetryData } from '../telemetry/models';
import { PanelContent } from './panelcontent';
import { PanelHeader } from './panelheader/panelheader';
import type { TitleChangeHandler } from './panelheader/panelheadertitle';
import { PanelResizer } from './panelResizer';
import type { CommonPanelProps } from './panelUtil';
import { PanelLocation, PanelScope, PanelSize } from './panelUtil';
import type { PanelNodeData } from './types';

export type PanelContainerProps = {
  noNodeSelected: boolean;
  panelScope: PanelScope;
  suppressDefaultNodeSelectFunctionality?: boolean;
  pivotDisabled?: boolean;
  readOnlyMode?: boolean;
  node: PanelNodeData | undefined;
  nodeHeaderItems: JSX.Element[];
  pinnedNode: PanelNodeData | undefined;
  pinnedNodeHeaderItems: JSX.Element[];
  layerProps?: ILayerProps;
  canResubmit?: boolean;
  overrideWidth?: string;
  resubmitOperation?: (nodeId: string) => void;
  onUnpinAction?: () => void;
  trackEvent(data: PageActionTelemetryData): void;
  toggleCollapse: () => void;
  onCommentChange: (nodeId: string, panelCommentChangeEvent?: string) => void;
  onTitleChange: TitleChangeHandler;
  onTitleBlur?: (prevTitle: string) => void;
  setOverrideWidth?: (width: string | undefined) => void;
  canShowLogicAppRun?: boolean;
  showLogicAppRun?: () => void;
} & CommonPanelProps;

export const PanelContainer = ({
  isCollapsed,
  panelLocation,
  noNodeSelected,
  panelScope,
  suppressDefaultNodeSelectFunctionality,
  canResubmit,
  resubmitOperation,
  onUnpinAction,
  readOnlyMode,
  node,
  nodeHeaderItems,
  pinnedNode,
  pinnedNodeHeaderItems,
  toggleCollapse,
  trackEvent,
  onCommentChange,
  onTitleChange,
  onTitleBlur,
  setOverrideWidth,
  overrideWidth,
  isResizeable,
  mountNode,
  canShowLogicAppRun,
  showLogicAppRun,
}: PanelContainerProps) => {
  const intl = useIntl();

  const selectedElementRef = useRef<HTMLElement | null>(null);

  const canResize = !!(isResizeable && setOverrideWidth);
  const isEmptyPane = noNodeSelected && panelScope === PanelScope.CardLevel;
  const isRight = panelLocation === PanelLocation.Right;
  const pinnedNodeId = pinnedNode?.nodeId;
  const pinnedNodeIfDifferent = pinnedNode && pinnedNode.nodeId !== node?.nodeId ? pinnedNode : undefined;

  const drawerWidth = isCollapsed
    ? PanelSize.Auto
    : (canResize ? overrideWidth : undefined) ?? (pinnedNodeIfDifferent ? PanelSize.DualView : PanelSize.Medium);

  useEffect(() => {
    selectedElementRef.current = node?.nodeId ? document.getElementById(`msla-node-${node.nodeId}`) : null;
  }, [node?.nodeId]);

  useEffect(() => {
    selectedElementRef.current?.focus();
  }, [isCollapsed]);

  const renderHeader = useCallback(
    (headerNode: PanelNodeData): JSX.Element => {
      const { nodeId } = headerNode;
      const panelHasPinnedNode = !!pinnedNodeIfDifferent;
      const isPinnedNode = pinnedNodeId === nodeId;
      const canUnpin = !!onUnpinAction && isPinnedNode;

      return (
        <PanelHeader
          nodeData={headerNode}
          isCollapsed={isCollapsed}
          isOutermostPanel={!panelHasPinnedNode || !isPinnedNode}
          headerItems={isPinnedNode ? pinnedNodeHeaderItems : nodeHeaderItems}
          headerLocation={panelLocation}
          noNodeSelected={noNodeSelected}
          panelScope={panelScope}
          suppressDefaultNodeSelectFunctionality={suppressDefaultNodeSelectFunctionality}
          readOnlyMode={readOnlyMode}
          canResubmit={canResubmit}
          canShowLogicAppRun={canShowLogicAppRun}
          showLogicAppRun={showLogicAppRun}
          onUnpinAction={canUnpin ? onUnpinAction : undefined}
          resubmitOperation={() => resubmitOperation?.(nodeId)}
          commentChange={(newValue) => onCommentChange(nodeId, newValue)}
          toggleCollapse={toggleCollapse}
          onTitleChange={onTitleChange}
          onTitleBlur={onTitleBlur}
        />
      );
    },
    [
      pinnedNodeIfDifferent,
      pinnedNodeId,
      onUnpinAction,
      isCollapsed,
      pinnedNodeHeaderItems,
      nodeHeaderItems,
      panelLocation,
      noNodeSelected,
      panelScope,
      suppressDefaultNodeSelectFunctionality,
      readOnlyMode,
      canResubmit,
      canShowLogicAppRun,
      showLogicAppRun,
      toggleCollapse,
      onTitleChange,
      onTitleBlur,
      resubmitOperation,
      onCommentChange,
    ]
  );

  const panelLabel = intl.formatMessage({
    defaultMessage: 'Operation details panel',
    id: 'nV2Spt',
    description: 'label for operation details panel component',
  });

  const panelErrorMessage = intl.formatMessage({
    defaultMessage: 'Error loading operation data',
    id: '62Ypnr',
    description: 'label for panel error',
  });

  const panelCollapseTitle = intl.formatMessage({
    defaultMessage: 'Collapse',
    id: 'lX30/R',
    description: 'Text of Tooltip to collapse',
  });

  const renderPanelContents = useCallback(
    (contentsNode: NonNullable<typeof node>, type: 'pinned' | 'selected'): JSX.Element => {
      const { errorMessage, isError, isLoading, nodeId, onSelectTab, selectedTab, tabs } = contentsNode;
      return (
        <div className={mergeClasses('msla-panel-layout', `msla-panel-layout-${type}`)}>
          {renderHeader(contentsNode)}
          <div className="msla-panel-contents">
            {isLoading ? (
              <div className="msla-loading-container">
                <Spinner size={'large'} />
              </div>
            ) : isError ? (
              <MessageBar messageBarType={MessageBarType.error}>{errorMessage ?? panelErrorMessage}</MessageBar>
            ) : (
              <PanelContent tabs={tabs} trackEvent={trackEvent} nodeId={nodeId} selectedTab={selectedTab} selectTab={onSelectTab} />
            )}
          </div>
        </div>
      );
    },
    [renderHeader, panelErrorMessage, trackEvent]
  );

  const minWidth = pinnedNode ? Number.parseInt(PanelSize.DualView, 10) : undefined;

  if (suppressDefaultNodeSelectFunctionality) {
    // Used in cases like BPT where we do not want to show the panel during node selection
    return null;
  }

  return (
    <OverlayDrawer
      aria-label={panelLabel}
      className="msla-panel-container"
      modalType="non-modal"
      mountNode={{
        className: 'msla-panel-host-container',
        element: mountNode,
      }}
      open={true}
      position={isRight ? 'end' : 'start'}
      style={{ position: 'absolute', width: drawerWidth }}
    >
      {isEmptyPane || isCollapsed ? (
        <Button
          appearance="subtle"
          aria-label={panelCollapseTitle}
          className={mergeClasses('collapse-toggle', isRight ? 'right' : 'left', isCollapsed && 'collapsed', 'empty')}
          icon={<ChevronDoubleRightFilled />}
          onClick={toggleCollapse}
          data-automation-id="msla-panel-header-collapse-nav"
        />
      ) : null}
      {isCollapsed ? null : (
        <>
          <div
            className={mergeClasses(
              'msla-panel-container-nested',
              `msla-panel-container-nested-${panelLocation.toLowerCase()}`,
              !isEmptyPane && pinnedNodeIfDifferent && 'msla-panel-container-nested-dual'
            )}
          >
            {isEmptyPane ? (
              <EmptyContent />
            ) : (
              <>
                {node ? renderPanelContents(node, 'selected') : null}
                {pinnedNodeIfDifferent ? (
                  <>
                    <Divider vertical={true} />
                    {renderPanelContents(pinnedNodeIfDifferent, 'pinned')}
                  </>
                ) : null}
              </>
            )}
          </div>
          {canResize ? <PanelResizer minWidth={minWidth} updatePanelWidth={setOverrideWidth} /> : null}
        </>
      )}
    </OverlayDrawer>
  );
};
