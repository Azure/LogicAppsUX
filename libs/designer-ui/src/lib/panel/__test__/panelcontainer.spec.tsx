import { PanelLocation, PanelScope } from '../panelUtil';
import type { PanelContainerProps } from '../panelcontainer';
import { PanelContainer } from '../panelcontainer';
import * as React from 'react';
import * as ReactShallowRenderer from 'react-test-renderer/shallow';
import { describe, vi, beforeEach, afterEach, beforeAll, afterAll, it, test, expect } from 'vitest';
describe('ui/workflowparameters/workflowparameter', () => {
  let minimal: PanelContainerProps, renderer: ReactShallowRenderer.ShallowRenderer;
  beforeEach(() => {
    minimal = {
      isCollapsed: false,
      panelLocation: PanelLocation.Right,
      noNodeSelected: false,
      panelScope: PanelScope.CardLevel,
      panelHeaderMenu: [],
      showCommentBox: true,
      tabs: {},
      title: 'test title',
      width: '630px',
      onCommentChange: vi.fn(),
      trackEvent: vi.fn(),
      setSelectedTab: vi.fn(),
      toggleCollapse: vi.fn(),
      onTitleChange: vi.fn(),
    };
    renderer = ReactShallowRenderer.createRenderer();
  });

  afterEach(() => {
    renderer.unmount();
  });

  it('should construct.', () => {
    const panel = renderer.render(<PanelContainer {...minimal} />);
    expect(panel).toMatchSnapshot();
  });

  it('should render.', () => {
    renderer.render(<PanelContainer {...minimal} />);
    const panel = renderer.getRenderOutput();

    expect(panel.props.className).toBe('msla-panel-container');
    expect(panel.props.headerClassName).toBe('msla-panel-header');
    expect(panel.props.headerText).toBe(minimal.title);
    expect(panel.props.customWidth).toBe(minimal.width);
  });
});
