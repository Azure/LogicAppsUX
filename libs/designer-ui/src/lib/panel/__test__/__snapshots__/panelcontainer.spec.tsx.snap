// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/workflowparameters/workflowparameter should construct. 1`] = `
<StyledPanelBase
  aria-label="panel"
  className="msla-panel-container"
  customWidth="630px"
  focusTrapZoneProps={
    Object {
      "disabled": false,
    }
  }
  hasCloseButton={false}
  headerClassName="msla-panel-header"
  headerText="test title"
  isBlocking={false}
  isOpen={true}
  onRenderHeader={[Function]}
  styles={
    Object {
      "content": Object {
        "padding": "1rem 2rem",
      },
      "main": Object {
        "overflow": "hidden",
      },
      "scrollableContent": Object {
        "height": "100%",
      },
    }
  }
  type={7}
>
  <React.Fragment>
    <PanelContent
      tabs={Object {}}
      trackEvent={[MockFunction]}
    />
  </React.Fragment>
</StyledPanelBase>
`;
